File: 20230920_Wednesday/20230920.rb
# Leetcode: 1658. Minimum Operations to Reduce X to Zero.
# https://leetcode.com/problems/minimum-operations-to-reduce-x-to-zero/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 203 ms, faster than 100.00% of Ruby online submissions for Minimum
# Operations to Reduce X to Zero.
# Memory Usage: 223.3 MB, less than 100.00% of Ruby online submissions for
# Minimum Operations to Reduce X to Zero.
# 2023.09.20 Daily Challenge.
# @param {Integer[]} nums
# @param {Integer} x
# @return {Integer}
def min_operations(nums, x)
  # Based on:
  # https://leetcode.com/problems/minimum-operations-to-reduce-x-to-zero/discuss/2136570/Change-Your-Perspective-or-JAVA-Explanation
  sum = nums.sum
  ml, cs = -1, 0 # ml - max lenght, cs - current sum.
  l = 0
  (0...nums.size).each do |r|
    cs += nums[r]
    while l <= r && (cs > (sum-x))
      cs -= nums[l]
      l += 1
    end
    ml = [ml, r-l+1].max if cs == sum - x
  end
  return -1 == ml ? -1 : nums.size - ml
end

File: 20230920_Wednesday/20230920.java
// https://leetcode.com/problems/minimum-operations-to-reduce-x-to-zero/discuss/2136570/Change-Your-Perspective-or-JAVA-Explanation
public int minOperations(int[] nums, int x) {
	int sum = 0;
	for (int num: nums) sum += num;

	int maxLength = -1, currSum = 0;
	for (int l=0, r=0; r<nums.length; r++) {
		currSum += nums[r];
		while (l <= r && currSum > sum - x) currSum -= nums[l++];
		if (currSum == sum - x) maxLength = Math.max(maxLength, r-l+1);
	}

	return maxLength == -1 ? -1 : nums.length - maxLength;
}
